Paper: Dask: Parallel Computation with Blocked algorithms and Task Scheduling (Rocklin)

mrocklin opened the pull request:
[Edit by jaberg] http://nipy.bic.berkeley.edu:5000/download/24

jaberg commented:
@katyhuff: I contacted a potential reviewer for this one.

mrocklin commented:
I would also be quite happy with reviews of the form "I don't get it" including points where confusion set in so that I could help to correct things.

katyhuff commented:
I'm planning to read it myself as well (probably today). @jaberg , let's assign ourselves to the one we want to do a review of.

Also, @mrocklin , please don't imagine we haven't already asked folks to review it. It's just that some reviewers have actually done the reviews that were requested and some didn't have the time. Between myself and whoever @jaberg thought of, we'll get you at least two reviews by the 30th.

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+-----------
+
+Normally humans write programs and then compilers/interpreters interpret them
+(e.g.  ``python``, ``javac``, ``clang``).  Sometimes humans disagree with how
```
I appreciate the humor in this: indeed I often find myself disagreeing quite forcefully with my compiler/interpreter!

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+------------
+
+The Scientific Python stack rarely leverages parallel computation.  Code built
+off of NumPy and Pandas generally runs in a single thread on data that fits
```
"Numpy and Pandas" -> "Numpy or Pandas" (or even "and/or"?)

mrocklin replied:
fixed

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
````diff
@@ -0,0 +1,730 @@
+task scheduling and policies to minimize memory footprint.  We then give two
+examples using ``dask.array`` on computational problems.  We then briefly
+discuss ``dask.bag`` and ``dask.dataframe``, two other collections in the
+``dask``` library.  We finish with thoughts about extension of this approach
````
There's an extra backtick here that I think wasn't intended.

mrocklin replied:
ah, good catch

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+using primitive Python dictionaries, tuples, and callables.  We use ``dask`` to
+create ``dask.array`` a parallel out-of-core NumPy clone using blocked
+algorithms.  This serves both as a general library for parallel out-of-core
+ndarrays but also as a demonstration that we can parallelize complex codebases
```
"ndarrays" should perhaps have backticks (as it's a class)?

mrocklin replied:
done

mrocklin replied:
and done throughout the document

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+Python data structures, namely dicts, tuples, functions, and arbitrary Python
+values.
+
+A **dask graph** is a dictionary mapping data-keys to values or tasks.
```
I found the phrase "data-keys" to be confusing.  I think I got it after stopping to consider it, but you might consider rewording to something giving a bit more explanation like "data-keys (akin to left-hand-side of an assignment operator)"

eteq replied:
Or it might also work to just make it clear that you'll explain that term more below

mrocklin replied:
changed a bit with your suggestion.  Please let me know if this is still unclear

eteq replied:
Much clearer now!

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+   (sum, ['x', (inc, 'x')])
+   (np.dot, np.array([...]), np.array([...]))
+
+To encode keyword arguments we recommend the use of ``functools.partial`` or
```
I don't quite understand the meaning of this sentence.  You're describing a specification (that's the name of the subsection, after all), but then you "recommend" the use of something that's necessary for the operation of a lot of code.  Are you saying the current implementation does not specify *how* keyword arguments work?  Or that it currently doesn't support keyword arguments?

mrocklin replied:
I now clarify that dask doesn't provide a space for keyword arguments but that in practice one can use partial.

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+all of your cores and operates on datasets that do not fit in memory.  It does
+this by building up a dask graph of blocked array algorithms.
+
+Several other large NumPy implementations exist, including Biggus_ an
```
I don't quite understand the purpose of this paragraph.  Are you saying these accomplish the same aim as Dask?  Or simply proof that it's ok to re-implement parts of NumPy?  So perhaps you should add a motivating sentence or two.

mrocklin replied:
At the end I mention that they all focus on slightly different things and say what dask.array's  motivations are

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+Dryad [Isa07]_ and Spark [Zah10]_.  These compute macroscopic operations with a
+collection of related in-memory operations.
+
+Dask.array takes a similar approach to linear algebra libraries but focuses
```
missing backticks on "dask.array" here?

mrocklin replied:
I tend not to use code highlighting when its the first word in the sentence.  Happy to change if you like.

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+*  Fancy indexing along single axes with lists or numpy arrays, ``x[:, [10, 1, 5]]``
+*  A variety of utility functions, ``bincount, where, ...``
+
+However dask.array is unable to handle any operation whose shape can not be
```
missing backticks on dark.array?

mrocklin replied:
Fixed.

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+Dynamic Task Scheduling
+-----------------------
+
+We now execute task graphs.  How we execute these graphs strongly impacts
```
Maybe instead "We now discuss how `dask` executes task graphs." or similar?

mrocklin replied:
Using your suggestion

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+task scheduling.  Additionally, data parallel systems like Dryad or Spark
+contain their own custom dynamic task schedulers.
+
+Traditional task scheduling with data dependencies scheduling literature
```
I think your point with this paragraph is that this is why you do *not* employ some pre-existing tool, is that right?  If so, you should say that (otherwise, a motivating sentence or two on why you're mentioning this would help)

mrocklin replied:
I now mention the criteria we had that no other scheduler met

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+worker from among the set of ready-to-run tasks.  This small choice governs the
+macro-scale performance of the scheduler.
+
+Instead for out-of-core computation we find value in choosing tasks that allow
```
I don't understand the use of "instead" here: instead of what?

mrocklin replied:
expanded

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+This lets us avoid spilling intermediate values to disk which hampers
+performance significantly.
+
+After several other policies we find that the policy of *last in, first out* is
```
I've lost the thread here: LIFO of *what*?  All of the tasks to be computed?  Or is this a subset that is "ready-to-run"?

mrocklin replied:
I've expanded on this a bit.  Please let me know if it is still unclear.

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+
+1.  Use ``concatenate`` and ``stack`` to manage large piles of HDF5 files (a
+    common case)
+2.  Using reductions and slicing to manipulate stacks of arrays
```
I'm confused by this (and still don't understand even after looking at the example): didn't you say above that dask can't handle slicing (because the size of output array is determined at compute time)?

mrocklin replied:
I clarify above that dask.array can handle slicing very well, just not slicing with another dask array of booleans

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+   show the average temperature difference between noon and midnight for year
+   2014
+
+This computation took about a minute on an old notebook computer.  The
```
Compared to the previous example, this paragraph seems incomplete.  Is this multi-threaded?  How did it compare to single-threaded?  Or is that the point of disk access: that it didn't change anything?  Need a bit more.

mrocklin replied:
I emphasize the point that this class of operations is often disk bound but that others, like image processing, are often compute bound and enjoy parallelism.

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+Python objects in dask.bag we are bound by the Global Interpreter Lock and so
+switch from using a multi-threaded scheduler to a multi-processing one.
+
+The ``dask.bag`` API contains functions like ``map`` and ``filter`` and
```
It's fine that this doesn't include full on "results", but perhaps you should add a bit about how this improves performance relative to not using `dask`?  Even just "it improved by a factor of a few" would be helpful.

mrocklin replied:
I mention that this provides the streaming/performance of cytoolz with multiple processes.

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+different blocks are well separated along an index.  For example in time series
+data we may know that all of January is in one block while all of February is
+in another.  Join, groupby, and range queries along this index are
+significantly faster when working on partitioned datasets.
```
Similar comment as for `bag`: even a very cursory single-sentence "results" would be helpful.

mrocklin replied:
I mention that this allows out-of-core and parallel execution when GIL is released.

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+   >>> x = da.ones((5000, 1000), chunks=(1000, 1000))
+   >>> u, s, v = da.svd(x)
+
+.. figure:: dask-svd.png
```
This is a really nice example - very impressive point!

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+   c = a.dot(b)  # another dask Array, not yet computed
+   c.store(out)  # Store result into output space
+
+**Results**: We do this same operation in different settings.
```
I want to  complement you on this section: very interesting and clearly-presented results!

mrocklin replied:
Thanks!

eteq commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,730 @@
+out-of-core execution.  These significantly extend the scale of data that is
+convenient to manipulate.
+
+**Low Barrier to Entry:** More importantly these collections demonstrate the
```
It's a bit confusing why this duplicates the ideas mentioned just one section up.  I think your point is that this is the "conclusions" so even if someone didn't read that they should read this, right?  If so, maybe just add a sentence further up noting this.

mrocklin replied:
Removed some of the redundancy here

eteq commented:
@mrocklin - I have reviewed this paper and in general I find it of high quality, very clear, well-organized, and complete.  (Also a very interesting project that I am excited to try out myself!)

I left a number of mostly stylistic (and hence you might call them "optional") comments inline.  Here I'm leaving some more high-level/general comments.

1. I saw no link to an actual website for the project.  Maybe that should be added in the intro?  Also useful would be to know the "status" of the package: is it considered "beta-level", or a full release, or what?  I saw it's on PyPI, but that's not quite enough to know for sure, especially if I'm considering using it for actual science or something.
2. Does `dask.array` work in any way with subclasses of ndarray? Or is this planned?  I'm not sure how "niche" that is, but certainly some projects do this and it would be useful to know if doing that rules out the use of `dask`.
3. I particularly noted the problem mentioned about how operations where the output array size is determined at runtime (i.e. the ``a[a>0]`` example) do not work with dask.  Is there a plan to address this by adding "breaks" between dask sets or anything like that?  Or will that be a "permanent" limitation? (I admit it's not clear if this is necessary for the paper... but I'm interested myself, so maybe others will be?)
4. Is `dask` compatible with other scheduler tools? That is, is it meant to be extended in that way *if the user wants*? I get the point about simplicity, but it would be useful to get a sense of how extensible it is for power users.
5. Is there any plan for having `dask.array` work in an "automated" fashion?  That is, you could imagine just running some function or context manager like ``daskify``, and then all numpy calls get converted to `dask.array` stuff. I imagine this is something that could get a lot of use, and I imagine a lot of readers would be interested in this.  Even if it's not in there right now, it would be good to say if this is planned so interested readers know what to watch out for.
6. The use of backticks around "dask.array" and similar usage is inconsistent: in some places (particularly near the beginning), it seems to be used uniformly, and later on it's often just raw text.  I don't particularly care which way you do it (Although maybe @katyhuff can say if there's a style rule/guideline for this?)... But it should be consistent, whichever way it is.

mrocklin commented:
Thank you for the thorough and thoughtful review @eteq. 

I believe I have touched on all of your comments in the recent commit.  To answer some of the questions in your final comment.

1.  Parts of dask are very solid, others less so.  I've added testament to the fact that it's "pretty stable and in daily use" in the document.  I wrote about this more thoroughly here http://matthewrocklin.com/blog/work/2015/05/19/State-of-Dask/
2.  No, subclassing `ndarray` can get weird.  We do implement the `__array__` though.
3.  We considered what you propose here but ultimately decided not to go ahead with it. https://github.com/ContinuumIO/dask/pull/260
4.  One can always make a scheduler (added in the document)
5.  No plans at the moment.  This would be interesting.  There was a theano project that did this with function decorators.  There is also the proposed `__numpy_ufunc__` protocol that would go a long way here if it ever gets released.

michaelpacer commented:
Hi @mrocklin just sitting down to review this — and given the time scale of response I figured sending any message if it was going to require any additional writing and not just editing was worth sending asap, even if it results in a scattered review. 

Your introduction is easy to follow but relies heavily on the notions of out-of-core and parallel computing to do the semantic work for you in expressing what you're doing that is novel and cool. But you do not explain what you mean by them, how they differ from each other, why you would want either, and what the challenges are for implementing them (in general). I'm knowledgeable enough that this didn't give me much pause, but am novice enough to parallel computing per se (in contrast to temporal scheduling problems and coherence management problems in general) that I needed to do side-research to look up exactly what you meant by out-of-core (which I had naïvely equated with parallelism). 

If you could spend a section following the Introduction wherein you explain what you mean by these two points, I believe that you'll greatly widen the number of people who can read this and fully understand it. And, fortunately, you have space to do this!

eteq commented:
@mrocklin - great, I think this has addressed all my comments.  Looks good to me at this point! (Although I concur with @michaelpacer that a bit more intro might be useful in those areas - it wasn't a problem for me, but I certainly see how it could be for others.)

eteq commented:
(So just to be clear, @katyhuff @jaberg, as far as I'm concerned this is ready, but I believe @michaelpacer has more comments coming)

lamblin commented:
I'm a bit late to the party, but in case it is useful, here is my (quick) review.
I find the article really interesting and well-written, the structure makes it clear what the goals were, how the approach works, then clearly details an example of how it was applied.
I was especially impressed by the simplicity of the data structures, the effectiveness of a simple scheduler, and the way the framework could be easily used by another team.

As @michaelpacer mentioned, the paper could benefit from an early introduction and definition of "out-of-core", "parallel computation", and "blocked algorithms" for those not so familiar with the field.

I would have a couple more questions that are more related to the implementation, but they may be out of the scope of the article:
1. Should the callable function of tasks be written in a special way to be able to take advantage of multi-thread parallelism, because of the GIL?
2. In particular, for `dask.array`, did you have to rewrite the callables (`arange`, `dot`, `dotmany`, `getitem`, ...) in a special way, or were you able to use `numpy` functions (or functions from elsewhere)?
3. How do inter-process or inter-machine communication work in the case of multi-process or distributed computation schedulers?
4. Do you (or could you) leverage the fact that some task could work in place, overwriting their input, for instance if they are the last task to need that data, in order to reduce memory usage?

I'll leave a couple of nit-picking comments as line comments if that is OK.

lamblin commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,777 @@
+The simplicity of dask graphs (no classes or frameworks) presents a very low
+barrier to entry.  Users only need to understand basic concepts common to
+Python (or indeed most modern languages) like dictionaries, tuples, and
+functions as variables.  As an example consider the work in [Tep15] in which
```
Missing link to the bibliography entry.

lamblin commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,777 @@
+briefly describe ``dask.bag`` and ``dask.dataframe``
+
+* ``dask.array`` = ``numpy`` + ``threading``
+* ``dask.bag = ``toolz`` + ``multiprocessing``
```
missing ending `` after "dask.bag"

lamblin commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,777 @@
+
+A *bag* is an unordered collection with repeats.  It is like a Python list but
+does not guarantee the order of elements.  Because we typically compute on
+Python objects in dask.bag we are bound by the Global Interpreter Lock and so
```
missing backticks around "dask.bag"?

lamblin commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,777 @@
+demonstrates the following:
+
+1.  How to interact with on-disk data
+2.  The blocked algorithms in dask.array achieve similar performance to modern
```
backticks around "dask.array"?

lamblin commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,777 @@
+2.  Use dask.array in a small amount of memory, pulling data from disk, using
+    four threads
+
+We different BLAS implementations
```
Seemingly missing word between "We" and "different".

lamblin commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,777 @@
+Example: Meteorology
+~~~~~~~~~~~~~~~~~~~~
+
+Performance is secondary to capability.  In this example we use dask.array to
```
backticks here as well?

michaelpacer commented:
In general i really liked this, and I agree that it's ready (though I do
think more aspects can be fleshed out). I have more comments that are
largely about parts that would bear more of an introduction that I'll put
together in a bit(ideally befoere the end of the night). But these are for
pedagogic/expository reasons rather than issues with content per se.

Apologies for the delays… some personal issues kept me away from my
computer and will be making it harder in the next few days but I figure
that comments will still be helpful(or at worst neutral) even if they are
after the deadline.

On Mon, Jun 29, 2015 at 9:23 PM, Erik Tollerud <notifications@github.com>
wrote:

> (So just to be clear, @katyhuff <https://github.com/katyhuff> @jaberg
> <https://github.com/jaberg>, as far as I'm concerned this is ready, but I
> believe @michaelpacer <https://github.com/michaelpacer> has more comments
> coming)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scipy-conference/scipy_proceedings/pull/132#issuecomment-116949970>
> .
>

lamblin commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,777 @@
+during dask's construction.
+
+Dask is used on a daily basis, both as a dependency in other projects in the
+SciPy ecosystem (xray, scikit-image, ...) and also in production in in private
```
duplicate "in": "in in private"

katyhuff commented:
This paper was extremely interesting and pleasant to read. I recommend accepting it. I have a few comments to potentially improve the paper in the sections below. Among those, the most important are:

- define what you mean "numpy clone" early on
- embiggen Figure 4
 
## Written Quality

This is an extremely readable narrative that captures some complex ideas in brief unpretentious language. A few minor suggestions follow:

- spell check (there are a couple of stray misspelled words in there... "lightwight")
- Reviewer's pet peeve: Comma use after prepositional and transitional phrases is inconsistent. Many instances lacking a comma would be made more readable with the addition of a comma. ("In these cases,"  "Commonly," "While less pleasant than our original code," "Also, in non-toy examples," "In practice," and similar.) Sorry to be the grammar police, but I do think it would be more readable if commas were added.
- What it means to be a "numpy clone" or numpy replacement should be stated simply and clearly in the text, as clone may be somewhat ambiguous for folks not familiar with this practice of cloning numpy (some people may not realize it's a reimplementation and stands alone. They may wonder "is it a subclass? " "is it a fork?")
- standardize whether you write NumPy or numpy. Similarly, Python or python.

## Technical Content

The work describes a piece of python software for approaching a parallel task scheduling -- a capability of potentially very broad interest in the scipy community.

## Novelty and Impact

The paper does an excellent job of noting prior work and motivating the advances made toward performance and flexibility that have not previously been possible with other tools.

## Verifiability

The software is described clearly and accompanied by open source code. Excellently described analysis that, in combination with the example workbooks referenced in the code documentation, is likely sufficiently detailed for reproduction.

## Other Requirements

**All mentioned software should be formally cited wherever possible.**
The author has done a good job of linking to noteworthy software, but actual citations are preferred where possible. Also,

- Flame and Plasma should be at least linked to, if not cited (because I've never heard of them and they're hard to google).
- Numpy and Pandas should be at least linked to, if not cited (because they are mentioned and are so key to the core of this effort)

### Length

The paper respects the 8-page limit, even if figure 4 were increased to a full page. 

### Figures

The figures are helpful.
- Figure 1 should be mentioned in the text where the example is explained (seems obvious, but because of the pagebreak, I almost missed the connection between the text and the figure)
- Figure 4 can (and should) be made larger with the `:figclass:w` attribute. Have you given that a shot  Alternatively, maybe `:figclass:p`.

### Code Snippets

**Code snippets should be formatted to fit inside a single column without overflow.**
A couple of code snippets overflow their boundaries. Please consider adjusting the displayed code with linebreaks to alleviate this. (e.g., shape and concatenate examples.)

katyhuff commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,777 @@
+Dynamic Task Scheduling
+-----------------------
+
+We now discuss how ``dask`` executes execute task graphs.  How we execute
```
I think "executes execute" is supposed to be, just, "executes" here.

katyhuff commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,777 @@
+
+Dynamic task scheduling has a rich literature and numerous projects, both
+within the Python ecosystem with projects like Spotify's Luigi_ for bulk data
+processing and projects without like DAGuE [Bos12]_ for more high performance
```
I had to re-read this sentence because of the "within the ecosystem like [...] and projects without like [...]" construction. It's a long sentence, but I think you can keep it if you maybe just say "without the python ecosystem like DAGuE..."

katyhuff commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,777 @@
+worker from among the set of ready-to-run tasks.  This small choice governs the
+macro-scale performance of the scheduler.
+
+Instead of these metrics commonly found in the literature we find that for
```
This might be more natural to read if it is "Instead of those metrics found..."

katyhuff commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,777 @@
+suited the needs of blocked algorithms for shared memory computation.  We
+needed a lightwight, easily installable Python solution that had latencies in
+the millisecond range and was mindful of memory use.  Traditional task
+scheduling with data dependencies scheduling literature usually focuses on
```
I feel like a word is missing.. 

maybe "data dependencies **in the** scheduling literature?"

katyhuff commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,777 @@
+
+Finally, power users can disregard the dask schedulers and create their own.
+Dask graphs are completely separate from the choice of scheduler and users may
+select the right scheduler for their class of problem or, if no ideal scheduler
```
FWIW, when I read this sentence I suddenly hoped for a couple of sentences about how to choose a scheduler. I think you sortof touch on this later in the paper a bit, but more implicitly than explicitly.

katyhuff commented on papers/matthew_rocklin/matthew_rocklin.rst:
```diff
@@ -0,0 +1,777 @@
+   +=======================+========+==============+
+   | Reference BLAS        | 6      |  18          |
+   +-----------------------+--------+--------------+
+   | OpenBLAS one thread   | 11     |  23          |
```
There's something kinda unpretty happening with this table (the Performance column labels are wrapped in a strange way.) I wonder if a wider first column could be forced? No big deal, but worth looking at if you have a chance.

mrocklin commented:
OK, I've handled a fair bit (though not yet all) of what was suggested.

1.  Various small edits made for obvious typos
2.  Added a section on modern hardware to describe parallel and out-of-core
3.  Added citations for those project that have academic references
4.  Embiggen figure 4
5.  Code snippet width

There are still some things to do.

mrocklin commented:
@lablin 

1. Should the callable function of tasks be written in a special way to be able to take advantage of multi-thread parallelism, because of the GIL?

If you use the treaded scheduler then yes, GIL releasing is good.  This is happening a lot in the community these days.  Projects seem to respond well to requests of the form "please release the GIL in these particular functions."  One can also use the multiprocessing scheduler for GIL holding computations.

2. In particular, for dask.array, did you have to rewrite the callables (arange, dot, dotmany, getitem, ...) in a special way, or were you able to use numpy functions (or functions from elsewhere)?

NumPy generally releases the GIL.  There were very few functions that required rewriting in NumPy.

3. How do inter-process or inter-machine communication work in the case of multi-process or distributed computation schedulers?

inter-process passes data through the main process, inter-machine communication communicates peer-to-peer.  See http://dask.pydata.org/en/latest/distributed.html .  

4. Do you (or could you) leverage the fact that some task could work in place, overwriting their input, for instance if they are the last task to need that data, in order to reduce memory usage?

One could.  Generally dask has not yet been used to tackle computations with computational constraints as interesting as Theano and so we haven't yet found motivation to do this sort of optimization.  It's also worth noting that the dask schedulers know nothing about arrays, they just execute python functions on python objects.

Note to self, mention and cite Theano.

mrocklin commented:
Some things I haven't done

*  Comma use after prepositional and transitional phrases is inconsistent.  I need to consider this more; it isn't natural to me.
*  Citations over links for all projects.  I've preferred citations over links when I can find them.  I can't find citations for all mentioned projects though.  Assistance welcome.
*  Figure internal references.  Haven't yet found the right prose to roll this in smoothly.

katyhuff commented:
@mrocklin : Thanks. I think you've successfully dealt with the key comments left by your reviewers. Any final edits need to be completed very soon in order to be included in the proceedings version of your paper. Please let me know when you have completed all of the edits that you intend to complete. As it stands, I'll mark your paper ready for merge.

mrocklin commented:
Turns out that I'm fairly swamped today.  My apologies.  Merge when ready.
